### PR TITLE
Experimental support for CRB 15000-12/1.27

### PR DESCRIPTION
**Pull request by mmmarcopalma**
Extending support for CRB 15000 to model 12-1.27 (GoFa 12 kg). 

Includes updates of: 

- `/config`, `/launch`, `/urdf` directories 
- .stl visual and collision mesh geometry 
   (link-3, link-5 and link-6 are referenced to `meshes/crb15000_5_95`)

### CODE REVIEW

**Comment by gavanderhoorn** [comment]
Thanks for the PR.

Some high-level comments:

- could you remove all the `desktop.ini` files please?
- could none of the meshes of the other variant be reused?
- please add a `roslaunch` test for the new variant
- please update the description in the manifest with the details about the new variant you added
- and if you'd like: please add yourself as an `author`, seeing as you added a new variant

**Comment by mmmarcopalma** [comment]
Thanks, I'll follow up on your comments.

> * could none of the meshes of the other variant be reused?

I could re-use 3+3 meshes (links 3-5-6). The ones I uploaded are exact copies of the existing ones.   

I'm not fluent yet in git, apologies for missing on some best-practice.

**Comment by gavanderhoorn** [comment]
> > * could none of the meshes of the other variant be reused?
> 
> I could re-use 3+3 meshes (links 3-5-6). The ones I uploaded are exact copies of the existing ones.

Ok, so this is perhaps due to me not being clear by what I meant by *reuse*.

A `xacro:macro` can refer to files 'anywhere', so you don't have to copy meshes if they are identical to files for other variants.

Just have your `<mesh filename="..."/>` point to the file you'd like to reuse from the other variant.

No need to copy anything.

**Comment by mmmarcopalma** [comment]
I added a commit with the requested changes. 
 
> * and if you'd like: please add yourself as an `author`, seeing as you added a new variant

Thanks, I added my name in the Package.xml

**Comment by mmmarcopalma** [comment]
Hello @gavanderhoorn, is there any news regarding this request? 
Thanks!

File: abb_crb15000_support/urdf/crb15000_12_127_macro.xacro [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    </link>
+    <!-- end of link list -->
+
+    <!-- joint list -->
+    <joint type="revolute" name="${prefix}joint_1">
+      <origin xyz="0 0 0.338" rpy="0 0 0"/>
+      <axis xyz="0 0 1"/>
+      <parent link="${prefix}base_link"/>
+      <child link="${prefix}link_1"/>
+      <limit effort="0" lower="${radians(-180)}" upper="${radians(180)}" velocity="${radians(125)}"/>
```
Comment (gavanderhoorn) [inline abb_crb15000_support/urdf/crb15000_12_127_macro.xacro]: Could you clarify which specific version of the `12/1.27` you're modelling here? According to the documentation I've found (`Product manual, CRB 15000`, `3HAC077389-001`, `Rev L`), there is an option for +-270 degrees as well:

image

File: abb_crb15000_support/urdf/crb15000_12_127_macro.xacro [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+      <parent link="${prefix}link_4"/>
+      <child link="${prefix}link_5"/>
+      <limit effort="0" lower="${radians(-180)}" upper="${radians(180)}" velocity="${radians(200)}"/>
+    </joint>
+    <joint type="revolute" name="${prefix}joint_6">
+      <origin xyz="0.101 0 0.080" rpy="0 0 0"/>
+      <axis xyz="1 0 0"/>
+      <parent link="${prefix}link_5"/>
+      <child link="${prefix}link_6"/>
+      <limit effort="0" lower="${radians(-180)}" upper="${radians(180)}" velocity="${radians(200)}"/>
```
Comment (gavanderhoorn) [inline abb_crb15000_support/urdf/crb15000_12_127_macro.xacro]: The position limits here seem to be different from the documentation I quoted earlier: `+-180` vs `+-270` degrees.

File: abb_crb15000_support/urdf/crb15000_12_127_macro.xacro (hunk 1 above)
Comment (gavanderhoorn) [reply abb_crb15000_support/urdf/crb15000_12_127_macro.xacro]: The joint velocity limit also seems off:

image

(this is from the CRB 15000 GoFa 10 & 12 Datasheet (from [here](https://new.abb.com/products/robotics/robots/collaborative-robots/crb-15000)))

Same for joints 2 & 3.

Could you clarify where you found the current limits?

**Review by gavanderhoorn — CHANGES_REQUESTED** [review]
Thanks again for the PR. Holiday period got in the way.

See the inline comments for some things that need to be checked and possibly corrected.

Apart from those, I'm wondering whether there is an issue with the normals of `visual` meshes `link_1.stl` and `link_4.stl`:

rviz_screenshot_2024_08_17-17_22_27

they appear to be inverted, at least partially.

File: abb_crb15000_support/urdf/crb15000_12_127_macro.xacro (hunk 1 above)
Comment (mmmarcopalma) [reply abb_crb15000_support/urdf/crb15000_12_127_macro.xacro]: these values were simply not updated from the GoFa 5, my bad.

Comment (mmmarcopalma) [reply abb_crb15000_support/urdf/crb15000_12_127_macro.xacro]: besides mesh normals, i realiize now link_4 is also not clsoed and has a bunch of non-manifold edges.  
fixing all of it right now.

File: abb_crb15000_support/urdf/crb15000_12_127_macro.xacro (hunk 2 above)
Comment (mmmarcopalma) [reply abb_crb15000_support/urdf/crb15000_12_127_macro.xacro]: see previous comment.

**Comment by mmmarcopalma** [comment]
Committed the requested changes. 
Thanks for supporting!

**Comment by mmmarcopalma** [comment]
Hello @gavanderhoorn, is there anything else I can do to push this PR? 
Thanks!

**Comment by gavanderhoorn** [comment]
I've re-enabled CI and added a couple of commits.

I believe this is in an OK state to merge. Thanks again for the PR @mmmarcopalma :+1: 

Let's wait on CI to turn green.

**Review by gavanderhoorn — APPROVED** [review]

**Comment by gavanderhoorn** [comment]
To prevent all the fixups from ending up in the commit history, I'll squash-merge this.

Provenance and attribution will be retained, of course.

**Comment by gavanderhoorn** [comment]
Thanks again @mmmarcopalma :+1:

**Comment by mmmarcopalma** [comment]
Great news, thanks for the support @gavanderhoorn, and thanks also to @gonzalocasas for initiating the process

**Pull request merged**